### PR TITLE
Refactor HLSL

### PR DESCRIPTION
**Pull request by huwb**
**Status**: WIP

This branch refactors shader code based on learnings from SRP versions.

### CODE REVIEW

**Comment by daleeidd** [comment]
What are the chances of us doing the clip surface refactor as part of this? It was simple enough to get it working in this branch. But I don't know how far away from finished this refactor is.

**Comment by huwb** [comment]
As for status, if memory serves, I think this might be mergeable now, it's done a good job of moving hlsl code.

The bigger thing was to actually change the functions like WorldToUV to match the srp versions. This would be great to do but is very error prone and will take some time to do.

As for your actual question specific to the clip thing, I'm not sure, but potentially yes.

Shall we get this branch merged? Could you try it on your side and of it looks good merge it? Maybe try Reimport the shaders folder when you do.

**Comment by huwb** [comment]
What is the clip surface refactor, is it related to getting it ready to work in hdrp? Sorry I'm a bit sleep deprived!

**Comment by daleeidd** [comment]
Sure. I can do that.

Sorry. I was a bit loose with the term. It was just having it work with this new structure which also HDRP uses.

**Comment by huwb** [comment]
Got it. You could include the changes into this branch before merging of they're logically related enough.

**Comment by daleeidd** [comment]
I brought over the clip surface stuff. The changes made to the clip surface shader in the other PR didn't work which is why it probably wasn't working in HDRP either. Got it working with the new files.

I also brought over some minor, but safe, changes not related to the clip surface from downstream.

I'll do another test tomorrow, but everything appears to be working fine. The only thing I can notice that we might want to address is the `CBUFFER`s have been mostly disabled in this refactor. I remember there being notes about Vulcan having issues with this stuff. Should this be addressed before merging?

**Comment by huwb** [comment]
I think it was just that CBUFFER is not defined in BIRP and it was never clear to me that constant buffers are supported. Although googling now, it looks like it is, for e.g. https://catlikecoding.com/unity/tutorials/scriptable-render-pipeline/custom-shaders/

I'm not sure how to get them in, should we try to define the macro if it isnt already? maybe we can find the definition in SRP.core or elsewhere.

If you want to look at this feel free, otherwise i will later.

The vulkan thing was just that every cbuffer relevant to a shader must have a unique name, otherwise they conflict.

**Comment by daleeidd** [comment]
Sorry. A little confused. That link is for SRP. Wrong link?

Constant Buffer macros are listed here:
https://docs.unity3d.com/Manual/SL-BuiltinMacros.html
But they don't work for compute shaders apparently. Is that what you're referring too?

I did find this comment about that (but not much else so far):
https://forum.unity.com/threads/cbuffer-and-compute-shader.164968/#post-3420530

**Comment by huwb** [comment]
Ok i think i just overlooked them before, not sure why. So lets just enable them.

If they don't work in compute we can find a solution.

**Comment by daleeidd** [comment]
Done. I have tested it on both Mac and Windows and they appear to function identically to master.

ffdce92 was interesting. I don't know why this affected shadows but not others.

I added some dummy macros for using constant buffers in compute to silence the errors. Let me know if you want a better filename/location.

**Comment by huwb** [comment]
Hm curious about the shadow lods, they all seem to work find for me in win64 editor. Don't want to drag this out unnecessarily, just understanding. ill anyway continue reviewing.

File: crest/Assets/Crest/Crest/Shaders/Macros.hlsl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Crest Ocean System
+
+// This file is subject to the MIT License as seen in the root of this folder structure (LICENSE)
+
+// MACROS - Macros to help with cross pipeline development
+
+#ifndef CREST_MACROS_H
+#define CREST_MACROS
+
+#ifndef CBUFFER_START
```
Comment (huwb) [inline crest/Assets/Crest/Crest/Shaders/Macros.hlsl]: So did you need to add a stub for this in the end - did the macro not work in 2018? this is a good solution and looks good, im just asking out of curiosity

**Comment by huwb** [comment]
This looks great, nice work.

I'm curious if any errors will come up on any compilers from the change of the type of slice index. However it didnt generate any for me when i reimported shaders and any remaining can come out in the wash.

Everything seemed to work for me, tested the example scenes in win64 editor and a standalone win64 build.

I'm happy to see this merged, merge it when you're ready. Thanks!

**Comment by daleeidd** [comment]
Here is what it looks like on Metal (also happens in Windows for me):

shadows issue

They look a little faint because the screenshot was taken at the first frame.

File: crest/Assets/Crest/Crest/Shaders/Macros.hlsl (hunk 1 above)
Comment (daleeidd) [reply crest/Assets/Crest/Crest/Shaders/Macros.hlsl]: Actually... I think I have done the wrong thing here. The macros aren't defined, but I didn't check if the stub is required. I've made an assumption. Also, there is a typo which I have fixed. I can copy the code from the PPv2 stack which defines these very macros for compute.

Comment (daleeidd) [reply crest/Assets/Crest/Crest/Shaders/Macros.hlsl]: I have implemented the stub from PPv2. Tested working fine on Metal. Thanks!

**Comment by daleeidd** [comment]
Might as well test the cbuffer macro Windows tomorrow. I will also merge if you have you haven't found anything else.

**Comment by huwb** [comment]
> Here is what the issue looked like on Metal (also happened on Windows for me)

Wow ok well nice find

Yes merge it in when you're happy

**Pull request merged**